lib/union_of.rb

```diff
@@ -0,0 +1,512 @@
+        else
+          super
+        end
+      end
+    end
+  end
+
+  module ReflectionExtension
+    def add_union_reflection(model, name, reflection)
+      model.union_reflections = model.union_reflections.merge(name.to_s => reflection)
+    end
+
+    private
+
+    def reflection_class_for(macro)
+      case macro
+      when :union_of
+        Reflection
+      else
+        super
+      end
+    end
+  end
+
+  module MacroReflectionExtension
+    def through_union_of? = false
+    def union_of?         = false
+  end
+
+  module RuntimeReflectionExtension
+    delegate :union_of?, :union_sources, to: :@reflection
+    delegate :name, :active_record_primary_key, to: :@reflection # FIXME(ezekg)
+  end
+
+  module ActiveRecordExtensions
+    extend ActiveSupport::Concern
+
+    included do
+      class_attribute :union_reflections, instance_writer: false, default: {}
+    end
+
+    class_methods do
+      def reflect_on_all_unions = union_reflections.values
+      def reflect_on_union(union)
+        union_reflections[union.to_s]
+      end
+    end
+  end
+
+  module ThroughReflectionExtension
+    delegate :union_of?, :union_sources, to: :source_reflection
+    delegate :join_scope, to: :source_reflection
+
+    def through_union_of? = through_reflection.union_of? || through_reflection.through_union_of?
+  end
+
+  module AssociationExtension
+    def scope
+      if reflection.union_of? || reflection.through_union_of?
+        Scope.create.scope(self)
+      else
+        super
+      end
+    end
+  end
+
+  module PreloaderExtension
+    def preloader_for(reflection)
+      if reflection.union_of?
+        Preloader::Association
+      else
+        super
+      end
+    end
+  end
+
+  module DelegationExtension
+    def delegated_classes
+      super << ReadonlyAssociationProxy
+    end
+  end
+
+  module JoinAssociationExtension
+    # Overloads Rails internals to prepend our left outer joins onto the join chain since Rails
+    # unfortunately does not do this for us (it can do inner joins via the LeadingJoin arel
+    # node, but it can't do outer joins because there is no LeadingOuterJoin node).
+    def join_constraints(foreign_table, foreign_klass, join_type, alias_tracker)
+      chain = reflection.chain.reverse
+      joins = super
+
+      # FIXME(ezekg) This is inefficient (we're recreating reflection scopes).
+      chain.zip(joins).each do |reflection, join|
+        klass = reflection.klass
+        table = join.left
+
+        if reflection.union_of?
+          scope = reflection.join_scope(table, foreign_table, foreign_klass, alias_tracker)
+          arel  = scope.arel(alias_tracker.aliases)
+
+          # Splice union dependencies, i.e. left joins, into the join chain. This is the least
+          # intrusive way of doing this, since we don't want to overload AR internals.
+          unless arel.join_sources.empty?
+            index = joins.index(join)
```

Comment:
How can we restructure these queries to be more efficient on the joined table? Since the joined table is more likely to have predicates added to it, we could swap the `FROM` and `JOIN` clauses to make it more efficient.

```sql
-- Slow
SELECT
  "licenses".*
FROM
  "licenses"
  LEFT OUTER JOIN "license_users" ON "license_users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
  AND "license_users"."license_id" = "licenses"."id"
  INNER JOIN "users" ON "users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
  AND (
    "users"."id" = "license_users"."user_id"
    OR "users"."id" = "licenses"."user_id"
  )
WHERE
  "licenses"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
  AND "users"."id" = 'a8c793b7-ef67-4866-9da0-0715e8ac76e4'
ORDER BY
  "licenses"."created_at" DESC;
```

```sql
-- Fast
SELECT
  "licenses".*
FROM
  "users"
  LEFT OUTER JOIN "license_users" ON "license_users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
  AND "license_users"."user_id" = "users"."id"
  INNER JOIN "licenses" ON "licenses"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
  AND (
    "licenses"."id" = "license_users"."license_id"
    OR "licenses"."user_id" = "users"."id"
  )
WHERE
  "users"."account_id" = 'f42ea6bb-e4fe-4e3d-873e-2fa22ebb2893'
  AND "users"."id" = 'a8c793b7-ef67-4866-9da0-0715e8ac76e4'
ORDER BY
  "licenses"."created_at" DESC;
```

Reply:
This might be something we have to do one-off, e.g. in `License.for_user` and `User.for_product`.

Reply:
I went the route of one-offs in 785c131c37fe2759d375de5a8adffceffea97574 because not all union queries needed to invert the from/join. I think this makes sense and doesn't necessarily signal a performance problem in `union_of`.